src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -0,0 +1,303 @@
+        pending_actors_.emplace(std::move(actor));
+      });
+
+  gcs_actor_scheduler_->SetScheduleSuccessHandler(
+      [this](std::shared_ptr<GcsActor> actor) {
+        auto worker_id = actor->GetWorkerID();
+        RAY_CHECK(!worker_id.IsNil());
+        RAY_CHECK(worker_to_created_actor_.emplace(worker_id, actor).second);
+
+        auto actor_id = actor->GetActorID();
+        auto node_id = actor->GetNodeID();
+        RAY_CHECK(!node_id.IsNil());
+        RAY_CHECK(node_to_created_actors_[node_id].emplace(actor_id, actor).second);
+
+        actor->UpdateState(rpc::ActorTableData::ALIVE);
+        actor->AsyncFlush();
+      });
+
+  RAY_LOG(INFO) << "Finished loading actors, size = " << registered_actors_.size();
+}
+
+GcsActorManager::GcsActorManager(boost::asio::io_context &io_context,
+                                 std::shared_ptr<gcs::RedisGcsClient> gcs_client,
+                                 gcs::GcsNodeManager &gcs_node_manager)
+    : GcsActorManager(io_context, std::move(gcs_client), gcs_node_manager,
+                      std::unique_ptr<gcs::GcsActorScheduler>(
+                          new gcs::GcsActorScheduler(io_context, gcs_node_manager))) {}
+
+void GcsActorManager::RegisterActor(
+    const ray::rpc::CreateActorRequest &request,
+    std::function<void(std::shared_ptr<GcsActor>)> &&callback) {
+  RAY_CHECK(callback);
+  const auto &actor_creation_task_spec = request.task_spec().actor_creation_task_spec();
+  auto actor_id = ActorID::FromBinary(actor_creation_task_spec.actor_id());
+
+  auto iter = registered_actors_.find(actor_id);
+  if (iter != registered_actors_.end()) {
+    // The actor already exists, just fast reply ok.
+    callback(iter->second);
+    return;
+  }
+
+  rpc::ActorTableData actor_table_data;
+  actor_table_data.set_actor_id(actor_id.Binary());
+  actor_table_data.set_job_id(request.task_spec().job_id());
+  actor_table_data.set_max_reconstructions(
+      actor_creation_task_spec.max_actor_reconstructions());
+  actor_table_data.set_remaining_reconstructions(
+      actor_creation_task_spec.max_actor_reconstructions());
+
+  auto dummy_object = TaskSpecification(request.task_spec()).ActorDummyObject().Binary();
+  actor_table_data.set_actor_creation_dummy_object_id(dummy_object);
+
+  actor_table_data.set_is_detached(actor_creation_task_spec.is_detached());
+  actor_table_data.mutable_owner_address()->CopyFrom(
+      request.task_spec().caller_address());
+
+  actor_table_data.set_state(ActorTableData::PENDING);
+  actor_table_data.mutable_task_spec()->CopyFrom(request.task_spec());
+
+  actor_table_data.mutable_address()->set_raylet_id(ClientID::Nil().Binary());
+  actor_table_data.mutable_address()->set_worker_id(WorkerID::Nil().Binary());
+
+  auto actor = std::make_shared<GcsActor>(std::move(actor_table_data), redis_gcs_client_);
+  actor->AsyncFlush([this, actor, callback] {
+    registered_actors_.emplace(actor->GetActorID(), actor);
```

Comment:
> Don't you have to add to the map before calling `AsyncFlush`? Or else, if you get a duplicate request, you will incorrectly flush the same actor twice.

Yes, there is a `registered_actors_` map to records all registered actors, it will filter the duplicate request at the beginning of the `GcsActorManager::RegisterActor`.

Reply:
I meant that you are adding to the `registered_actors_` *after* the `AsyncFlush` returns. So if a duplicate call to `RegisterActor` comes while the first one is still in flight, wouldn't this incorrectly flush the same actor twice?

This is a good example of something that should be covered by the unit tests. Can you add a regression test that catches this?

Reply:
> I meant that you are adding to the `registered_actors_` _after_ the `AsyncFlush` returns. So if a duplicate call to `RegisterActor` comes while the first one is still in flight, wouldn't this incorrectly flush the same actor twice?
> 
> This is a good example of something that should be covered by the unit tests. Can you add a regression test that catches this?

Thanks! I'm sure it's a bug, but I can't simply insert the actor to register_actors_ before it is flushed, because it will lead to a quick reply of the duplicated request, and at this time, it's possible that the actor hasn't flushed successfully.

I will fix it and add unit test, Thanks again.